Lib/typing.py

```diff
@@ -2726,11 +2726,26 @@ def __new__(cls, typename, bases, ns):
             class_getitem = _generic_class_getitem
             nm_tpl.__class_getitem__ = classmethod(class_getitem)
         # update from user namespace without overriding special namedtuple attributes
-        for key in ns:
+        for key, val in ns.items():
             if key in _prohibited:
                 raise AttributeError("Cannot overwrite NamedTuple attribute " + key)
-            elif key not in _special and key not in nm_tpl._fields:
-                setattr(nm_tpl, key, ns[key])
+            elif key not in _special:
+                if key not in nm_tpl._fields:
+                    setattr(nm_tpl, key, val)
+                try:
+                    set_name = type(val).__set_name__
+                except Exception:
```

Comment:
Worth commenting here that this matches the behavior for builtin classes?

Reply:
Added an essay for you in https://github.com/python/cpython/pull/111876/commits/c2c65c681dedae1446fcbfb840a274f915411b87 :-)